### PR TITLE
C++ port: add C++ equivalent of FSTDocumentKey.

### PR DESCRIPTION
**Pull request by var-const**
Also move kDocumentKeyPath to the only point of usage - make it a static
member variable of FieldPath.

### CODE REVIEW

**Comment by var-const** [comment]
Hi Rich,

Let's try out what we discussed today re. the review process. This PR is based on another pending PR I'm working on, #749.

Thanks,
Costa

File: Firestore/core/src/firebase/firestore/model/document_key.h [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+   * Creates and returns a new document key using '/' to split the string into
+   * segments.
+   */
+  static DocumentKey FromPathString(absl::string_view path) {
+    return DocumentKey{ResourcePath::FromString(path)};
+  }
+  /**
+   * Creates and returns a new document key with the given segments.
+   */
+  static DocumentKey FromSegments(std::initializer_list<std::string> list) {
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/model/document_key.h]: This is to be more similar to other SDKs, but we don't really need it. Because ResourcePath's constructor taking initializer list is not explicit, it's just as simple to do `DocumentKey({"a", "b", "c"})` as `DocumentKey::FromSegments({"a", "b", "c"})`.

File: Firestore/core/src/firebase/firestore/model/field_path.cc [hunk 2]
```diff
@@ -30,9 +30,6 @@ namespace model {
 
 namespace {
 
-// TODO(varconst): move to C++ equivalent of FSTDocumentKey.{h,cc}
-const char* const kDocumentKeyPath = "__name__";
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/model/field_path.cc]: Looking at other SDKs:

* each SDK has its own place to store this constant;
* this constant is only used in FieldPath and inside tests.

So it seemed easiest to just move it to the point of usage. Please let me know if there's a reason to have it inside DocumentKey/make it a global variable/etc.

File: Firestore/core/src/firebase/firestore/model/resource_path.cc [hunk 3]
```diff
@@ -27,7 +27,7 @@ namespace firebase {
 namespace firestore {
 namespace model {
 
-ResourcePath ResourcePath::Parse(const absl::string_view path) {
+ResourcePath ResourcePath::FromString(const absl::string_view path) {
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/model/resource_path.cc]: This is to be more similar to other SDKs.

File: Firestore/core/test/firebase/firestore/model/document_key_test.cc [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+  EXPECT_FALSE(key_from_moved_path.path().empty());
+  EXPECT_EQ(key_from_path_copy.path(), key_from_moved_path.path());
+}
+
+TEST(DocumentKey, CopyAndMove) {
+  DocumentKey key({"rooms", "firestore", "messages", "1"});
+  const std::string path_string = "rooms/firestore/messages/1";
+  EXPECT_EQ(path_string, key.path().CanonicalString());
+
+  auto copied = key;
```
Comment (rsgowman) [inline Firestore/core/test/firebase/firestore/model/document_key_test.cc]: style nit: It's best to avoid auto when its unclear what type is in use. But that statement leaves a lot up to the reader; after all, the type is just a few lines above. Personally, I typically only use auto when the type (more-or-less) otherwise appears on the same line, so I'd be inclined to not use auto here. But I don't much mind either way.

Optional.

(I'll refrain from pointing out subsequent instances of this. If you feel like adding the type here, then do so throughout.)

File: Firestore/core/test/firebase/firestore/model/document_key_test.cc [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+  ASSERT_DEATH_IF_SUPPORTED(DocumentKey::FromSegments({"foo"}), "");
+  ASSERT_DEATH_IF_SUPPORTED(DocumentKey::FromSegments({"foo", "bar", "baz"}),
+                            "");
+
+  ASSERT_DEATH_IF_SUPPORTED(DocumentKey::FromPathString(""), "");
+  ASSERT_DEATH_IF_SUPPORTED(DocumentKey::FromPathString("invalid//string"), "");
+  ASSERT_DEATH_IF_SUPPORTED(DocumentKey::FromPathString("invalid/key/path"),
+                            "");
+}
```
Comment (rsgowman) [inline Firestore/core/test/firebase/firestore/model/document_key_test.cc]: Consider adding ADIF(DocKey:FromPathStr("invalid"), "");

File: Firestore/core/test/firebase/firestore/model/document_key_test.cc [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+TEST(DocumentKey, IsDocumentKey) {
+  EXPECT_TRUE(DocumentKey::IsDocumentKey({}));
```
Comment (rsgowman) [inline Firestore/core/test/firebase/firestore/model/document_key_test.cc]: Curious. Do the other apis also consider an empty path to be a document key?

File: Firestore/core/test/firebase/firestore/model/document_key_test.cc [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+  EXPECT_TRUE(a < b);
+  EXPECT_TRUE(a <= b);
+  EXPECT_TRUE(b > a);
+  EXPECT_TRUE(b >= a);
+
+  EXPECT_TRUE(a < ab);
+  EXPECT_TRUE(a <= ab);
+  EXPECT_TRUE(ab > a);
+  EXPECT_TRUE(ab >= a);
+}
```
Comment (rsgowman) [inline Firestore/core/test/firebase/firestore/model/document_key_test.cc]: Consider adding comparison tests for empty with empty.
EXPECT_FALSE(empty < empty);
EXPECT_TRUE(empty <= empty);

^ assuming that's the behaviour we want. (Sometimes nullish things are defined to not compare equal even to themselves. I don't think we want that here, but it's mildly useful to explicitly call that out via a test.)

File: Firestore/core/test/firebase/firestore/model/document_key_test.cc [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+
```
Comment (rsgowman) [inline Firestore/core/test/firebase/firestore/model/document_key_test.cc]: Consider adding tests for [in]equality

EXPECT_TRUE(a == a);
EXPECT_FALSE(a != a);
etc.

File: Firestore/core/test/firebase/firestore/model/document_key_test.cc (hunk 4 above)
Comment (var-const) [reply Firestore/core/test/firebase/firestore/model/document_key_test.cc]: Done.

I generally prefer `auto` when copying/moving if the "original" is in scope, because the type can only be the type of the "original", and also because it makes it clear that there is no conversion involved. But I don't feel strongly about it, so changed to explicit typing.

File: Firestore/core/test/firebase/firestore/model/document_key_test.cc (hunk 5 above)
Comment (var-const) [reply Firestore/core/test/firebase/firestore/model/document_key_test.cc]: Done.

File: Firestore/core/test/firebase/firestore/model/document_key_test.cc (hunk 6 above)
Comment (var-const) [reply Firestore/core/test/firebase/firestore/model/document_key_test.cc]: They must, because they all implement the check as simple `path.size() % 2 == 0`. I'm not sure if they should, though, which is partially why I added this test. @wilhuff Gil, is this intentional?

File: Firestore/core/test/firebase/firestore/model/document_key_test.cc (hunk 7 above)
Comment (var-const) [reply Firestore/core/test/firebase/firestore/model/document_key_test.cc]: Done. Existing SDKs don't appear to treat empty specially.

File: Firestore/core/src/firebase/firestore/model/field_path.cc (hunk 2 above)
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/model/field_path.cc]: Feel free to make organizational changes but when you notice inconsistencies like these please fix them :-).

File: Firestore/core/test/firebase/firestore/model/document_key_test.cc (hunk 8 above)
Comment (var-const) [reply Firestore/core/test/firebase/firestore/model/document_key_test.cc]: Done. Added just these two, though, to make sure comparison to self is sensible. `EXPECT_EQ/NE` macros exercise those operators anyway.

File: Firestore/core/test/firebase/firestore/model/document_key_test.cc (hunk 4 above)
Comment (wilhuff) [reply Firestore/core/test/firebase/firestore/model/document_key_test.cc]: The other case where auto is preferred is when the type is onerous or noisy (e.g. `std::vector<MyType>::const_iterator`).

This usage is just barely on the other side of what would occur to me to do but I don't care strongly.

File: Firestore/core/test/firebase/firestore/model/document_key_test.cc (hunk 6 above)
Comment (wilhuff) [reply Firestore/core/test/firebase/firestore/model/document_key_test.cc]: It's valid though hard to construct.

Document keys and collection keys differ only in even vs odd numbers of segments. Accordingly the zero segment key is a document key, though it's one you can neither read nor write.

File: Firestore/core/src/firebase/firestore/model/document_key.h [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+#include <string>
+
+#include "Firestore/core/src/firebase/firestore/model/resource_path.h"
+#include "absl/strings/string_view.h"
+
+namespace firebase {
+namespace firestore {
+namespace model {
+
+class DocumentKey {
```
[the remark this answers is not in the document]
Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/model/document_key.h]: AFAICT, it's not at all against the style guide. From the examples listed, the style guide usually omits blank lines after methods that are entirely contained on a single line, but adds them when the definition spans lines. (See guide for examples). But I don't think there's any requirement here.

I'm fine with it as is.

File: Firestore/core/src/firebase/firestore/model/document_key.h [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+  static bool IsDocumentKey(const ResourcePath& path) {
+    return path.size() % 2 == 0;
+  }
+
+  const ResourcePath& path() const {
+    return path_;
+  }
+
+ private:
+  ResourcePath path_;
```
Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/model/document_key.h]: DocumentKeys are frequently used as keys in maps and extracted from the documents they name. Given that these are not that cheap to create (they contain a vector of strings) I wonder if we should make DocumentKey a pointer-hiding value type that contains a `std::shared_ptr<ResourcePath>` instead of just a `ResourcePath`.

This would make DocumentKeys substantially cheaper to pass around and would avoid is making maps like this:

std::unordered_map<std::shared_ptr<Document_key>, std::shared_ptr<Document>>

Comment (var-const) [reply Firestore/core/src/firebase/firestore/model/document_key.h]: What's the bigger concern, copying keys around or looking them up in a map? For lookup, wouldn't shared pointer make it slower because of an additional indirection (or not affect the time, in best case)?

If lookup is the bigger concern, then perhaps we should cache the hash (since DocumentKey is *mostly* immutable, it shouldn't change often)? If copying is the bigger concern, then shared pointer sounds good. Let me look more at the usages and get back to you.

Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/model/document_key.h]: Sorry, the concern I was trying to raise was copying.

If you look through the LocalStore we do a whole bunch of operations on documents track the affected keys. Similarly we compute views using sets of affected keys. All of these come from the underlying documents.

The `FSTDocumentKey` we're replacing, as an Objective-C object, is always heap allocated and always effectively tracked through shared pointers (that's what a strong reference under ARC effectively is). This means that making `DocumentKey` contain a `shared_ptr` we'll end up essentially no worse off that we were before.

Comment (var-const) [reply Firestore/core/src/firebase/firestore/model/document_key.h]: This makes shared pointer sound very reasonable. Is hashing of no concern at all (the two aren't mutually exclusive). Also, is FSTDocumentKey passed between threads?

Comment (var-const) [reply Firestore/core/src/firebase/firestore/model/document_key.h]: Just to throw an idea around, shared_ptr optimization could be moved to BasePath as well. That will make it affect the public FieldPath, too, for better or worse.

**Comment by wilhuff** [comment]
Unlike keys I don’t think field paths get shared like this so I’m not sure
it’s worth it. Keys are an obvious win case.

We can always refactor later if we find field paths are getting copied
around.

File: Firestore/core/src/firebase/firestore/model/document_key.h (hunk 10 above)
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/model/document_key.h]: FSTDocumentKey (by virtue of being part of FSTDocument) is passed between threads when we deliver snapshots to users. This is part of why I'm insistent that these structures be (mostly) immutable.

Unlike keys I don’t think field paths get shared like this so I’m not sure it’s worth it. Keys are an obvious win case.

We can always refactor later if we find field paths are getting copied around.

File: Firestore/core/src/firebase/firestore/model/field_path.cc (hunk 2 above)
Comment (var-const) [reply Firestore/core/src/firebase/firestore/model/field_path.cc]: To be fair, there seem to be reasons for this inconsistencies. E.g., Java can't have a global variable, so the variable is wrapped into a class, whereas Typescript even has a nice syntax for exporting variables.

File: Firestore/core/src/firebase/firestore/model/document_key.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (var-const) [reply Firestore/core/src/firebase/firestore/model/document_key.h]: Done.

File: Firestore/core/src/firebase/firestore/model/document_key.h [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+  static DocumentKey FromSegments(std::initializer_list<std::string> list) {
+    return DocumentKey{ResourcePath{list}};
+  }
+
+  /**
+   * Returns a shared instance of an empty document key.
+   */
+  static const DocumentKey& Empty();
+
+  static bool IsDocumentKey(const ResourcePath& path) {
```
[the remark this answers is not in the document]
Comment (var-const) [reply Firestore/core/src/firebase/firestore/model/document_key.h]: Done.

File: Firestore/core/src/firebase/firestore/model/document_key.h (hunk 10 above)
Comment (var-const) [reply Firestore/core/src/firebase/firestore/model/document_key.h]: Done.

Note that it means that DocumentKey may now be in an invalid moved-from state. To make the behavior match the previous iteration better, I did:

```
const ResourcePath& DocumentKey::path() const {
  return path_ ? *path_ : Empty().path();
}
```

which effectively guarantees that moved-from state is always valid.

File: Firestore/core/src/firebase/firestore/model/document_key.cc [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+DocumentKey DocumentKey::FromPathString(const absl::string_view path) {
+  return DocumentKey{ResourcePath::FromString(path)};
+}
+
+DocumentKey DocumentKey::FromSegments(std::initializer_list<std::string> list) {
+  return DocumentKey{ResourcePath{list}};
+}
+
+const DocumentKey& DocumentKey::Empty() {
+  static DocumentKey empty;
```
Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/model/document_key.cc]: Should this be `static const DocumentKey empty`?

File: Firestore/core/src/firebase/firestore/model/document_key.h [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+
+  /** Creates and returns a new document key with the given segments. */
+  static DocumentKey FromSegments(std::initializer_list<std::string> list);
+
+  /** Returns a shared instance of an empty document key. */
+  static const DocumentKey& Empty();
+
+  /** Returns true iff the given path is a path to a document. */
+  static bool IsDocumentKey(const ResourcePath& path);
+  /** The path to the document. */
```
Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/model/document_key.h]: missing newline?

File: Firestore/core/src/firebase/firestore/model/document_key.h [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+#include <memory>
+#include <string>
+
+#include "absl/strings/string_view.h"
+
+namespace firebase {
+namespace firestore {
+namespace model {
+
+class ResourcePath;
```
Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/model/document_key.h]: I'm not sure this build-time optimization is worthwhile. It forces the entire implementation out into the .cc file, where this would otherwise be a very lightweight wrapper around a `ResourcePath`.

I don't feel strongly that these things should be defined inline and perhaps this is beneficial because it forces all these things to be defined out-of-line.

File: Firestore/core/src/firebase/firestore/model/document_key.cc (hunk 12 above)
Comment (var-const) [reply Firestore/core/src/firebase/firestore/model/document_key.cc]: Done.

File: Firestore/core/src/firebase/firestore/model/document_key.h (hunk 14 above)
Comment (var-const) [reply Firestore/core/src/firebase/firestore/model/document_key.h]: Reverted back.

File: Firestore/core/src/firebase/firestore/model/document_key.h (hunk 13 above)
Comment (var-const) [reply Firestore/core/src/firebase/firestore/model/document_key.h]: Done.
